homeassistant/components/roomba.py

```diff
@@ -0,0 +1,152 @@
+}
+
+SERVICE_CLEAN = 'clean'
+SERVICE_DOCK = 'dock'
+SERVICE_PAUSE = 'pause'
+
+
+def setup(hass, config):
+    """Set up the Roomba component."""
+    from roomba import Roomba
+
+    roomba = Roomba(
+        address=config[DOMAIN][CONF_HOSTNAME],
+        blid=config[DOMAIN][CONF_USERNAME],
+        password=config[DOMAIN][CONF_PASSWORD],
+        cert_name=config[DOMAIN][CONF_CERT],
+        continuous=config[DOMAIN][CONF_CONTINUOUS]
+    )
+    roomba_hub = RoombaHub(hass, config[DOMAIN], roomba)
+    hass.data[ROOMBA_ROBOTS] = [roomba_hub]
+
+    if not roomba_hub.login():
+        _LOGGER.debug('Failed to communicate with Roomba')
+        return False
+
+    # Request data update
+    roomba_hub.update()
+
+    # Setup sensors and switches
+    for component in ['sensor']:
+        discovery.load_platform(hass, component, DOMAIN, {}, config)
+
+    def handle_clean(call):
+        """Start cleaning cycle."""
+        roomba.send_command('clean')
+
+    def handle_dock(call):
+        """Return to base."""
+        roomba.send_command('dock')
+
+    def handle_pause(call):
+        """Pause cleaning cycle."""
+        roomba.send_command('pause')
+
+    hass.services.register(DOMAIN, SERVICE_CLEAN, handle_clean)
+    hass.services.register(DOMAIN, SERVICE_DOCK, handle_dock)
+    hass.services.register(DOMAIN, SERVICE_PAUSE, handle_pause)
+    return True
+
+
+class RoombaTimeoutError(Exception):
+    """Roomba Timeout occurred."""
+
+    pass
+
+
+class RoombaHub(object):
+    """A Roomba wrapper class."""
+
+    def __init__(self, hass, config, roomba):
+        """Initialize the Roomba hub."""
+        self.config = config
+        self._roomba = roomba
+        self._hass = hass
+        self.data = None
+
+    def _wait_for_update(self, timeout=10):
+        from time import sleep
+        tries = 0
+        while not self._roomba.master_state:
+            _LOGGER.debug('No data received from Roomba yet')
+            sleep(1)
+            if tries > timeout:
+                raise RoombaTimeoutError(
+                    'No data received after {} seconds'.format(timeout))
+            tries = tries + 1
+        _LOGGER.debug('Communication with roomba succeeded')
+
+    def login(self):
+        """Login to Roomba."""
+        try:
+            _LOGGER.debug('Trying to connect to Roomba')
+            self._roomba.connect()
+            self._wait_for_update()
+            _LOGGER.debug('Roomba state: %s',
+                          self._roomba.cleanMissionStatus_phase)
+            return True
+        except RoombaTimeoutError:
+            _LOGGER.error('Unable to connect to Roomba')
+            return False
+
+    @Throttle(timedelta(seconds=5))
+    def update(self):
+        """Reconnect to Roomba to request data update."""
+        from time import sleep
+        _LOGGER.debug('Running Roomba update %s',
+                      self._hass.data[ROOMBA_ROBOTS])
+        # Wait a few seconds for the data
+        sleep(3)
```

Comment:
You can not use sleep inside home-assitant. Look into `helpers/events.py` for time handlings